packages/@ember/-internals/metal/lib/array.ts

```diff
@@ -58,82 +56,49 @@ export function replaceInNativeArray<T>(
 }
 
 interface ArrayObserverOptions {
-  willChange?: string;
-  didChange?: string;
+  willChange: string;
+  didChange: string;
 }
 
-type Operation = (
-  obj: ObjectHasArrayObservers,
+type Operation<T> = (
+  obj: ObservedObject<T>,
   eventName: string,
   target: object | Function | null,
   callbackName: string
 ) => void;
 
-function arrayObserversHelper(
-  obj: ObjectHasArrayObservers,
+function arrayObserversHelper<T>(
+  obj: ObservedObject<T>,
   target: object | Function | null,
-  opts: ArrayObserverOptions | undefined,
-  operation: Operation,
-  notify: boolean
-): ObjectHasArrayObservers {
-  let willChange = (opts && opts.willChange) || 'arrayWillChange';
-  let didChange = (opts && opts.didChange) || 'arrayDidChange';
-  let hasObservers = obj.hasArrayObservers;
+  opts: ArrayObserverOptions,
+  operation: Operation<T>
+): ObservedObject<T> {
+  let { willChange, didChange } = opts;
 
   operation(obj, '@array:before', target, willChange);
   operation(obj, '@array:change', target, didChange);
 
-  if (hasObservers === notify) {
-    notifyPropertyChange(obj, 'hasArrayObservers');
-  }
+  /*
+   * Array proxies have a `_revalidate` method which must be called to set
+   * up their internal array observation systems.
+   */
+  obj._revalidate?.();
```

Comment:
In this patch `hasArrayObserver` has been removed. That property was not directly deprecated in 3.x, however we've removed a number of APIs in this patch which were related to `someArray.addArrayObserver` (APIs which were only meaningful once you had an observer installed).

Because it is removed, it no longer needs to be notified of change. However the `_revalidate` hook on `ArrayProxy` instances was being called via `PROPERTY_DID_CHANGE` symbol notification. Without trying to rewrite that system we have two options: Either call `_revalidate()` directly, or keep the property `hasArrayObserver` around just to trigger the `PROPERTY_DID_CHANGE`. Here I've chosen to call `_revalidate()` directly.